services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/models/shared_store.py

```diff
@@ -1,58 +1,101 @@
 
 from ..core.settings import ApplicationSettings
 
-ContainerNameStr = str
+ContainerNameStr: TypeAlias = str
 
 STORE_FILE_NAME: Final[str] = "data.json"
 
 
-class SharedStore(BaseModel):
-    _shared_store_dir: Path = PrivateAttr()
+class _StoreMixin(BaseModel):
+    _shared_store_dir: Path | None = PrivateAttr()
+    _persist_lock: Lock | None = PrivateAttr()
 
-    compose_spec: Optional[str] = Field(
+    async def __aenter__(self) -> None:
+        assert self._persist_lock  # nosec
+        await self._persist_lock.acquire()
+        return None
+
+    async def __aexit__(self, *args) -> None:
+        await self._persist_to_disk()
+
+        assert self._persist_lock  # nosec
+        self._persist_lock.release()
+
+    async def _persist_to_disk(self) -> None:
+        assert self._shared_store_dir  # nosec
+        async with aiofiles.open(
+            self._shared_store_dir / STORE_FILE_NAME, "w"
+        ) as data_file:
+            await data_file.write(self.json())
+
+
+class SharedStore(_StoreMixin):
+    """
+    When used as a context manager will persist the state to the disk upon exit.
+
+    NOTE: when updating the contents of the shared store always use a context manger
+    to avoid concurrency issues.
+
+    Example:
+        async with shared_store:
+            copied_list = deepcopy(shared_store.container_names)
+            copied_list.append("a_container_name")
+            shared_store.container_names = copied_list
+    """
+
+    compose_spec: str | None = Field(
         default=None, description="stores the stringified compose spec"
     )
     container_names: list[ContainerNameStr] = Field(
         default_factory=list,
         description="stores the container names from the compose_spec",
     )
 
-    # NOTE: setting up getter and setter does not work.
-    def set_shared_store_dir(self, shared_store_dir: Path) -> None:
-        self._shared_store_dir = shared_store_dir
+    volume_states: dict[VolumeCategory, VolumeState] = Field(
+        default_factory=dict, description="persist the state of each volume"
+    )
 
-    async def clear(self):
-        self.compose_spec = None
-        self.container_names = []
-        await self.persist_to_disk()
+    async def _setup_initial_volume_states(self) -> None:
+        async with self:
+            for category, status in [
+                (VolumeCategory.INPUTS, VolumeStatus.CONTENT_NO_SAVE_REQUIRED),
+                (VolumeCategory.SHARED_STORE, VolumeStatus.CONTENT_NO_SAVE_REQUIRED),
+                (VolumeCategory.OUTPUTS, VolumeStatus.CONTENT_NEEDS_TO_BE_SAVED),
+                (VolumeCategory.STATES, VolumeStatus.CONTENT_NEEDS_TO_BE_SAVED),
+            ]:
+                self.volume_states[category] = VolumeState(status=status)
 
     @classmethod
     async def init_from_disk(cls, shared_store_dir: Path) -> "SharedStore":
+        def _init_private(obj: SharedStore):
+            # pylint: disable=protected-access
```

Comment:
do you really need that pylint disable here? can't you pass them through the constructor?

Reply:
Fields marked as `PrivateAttr` cannot be initialised via the constructor. Also they are not touched by Pedantic and are left to the user to mange their internals during the lifecycle of the object.

I did a bit of refactoring. Let me know if it looks better now.